### PR TITLE
added storage delegation feature

### PR DESCRIPTION
**Pull request by andreainnocenti**
### Release required?

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [x] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

### CODE REVIEW

File: storage_delegation/variables.tf [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  sensitive   = true
+  type        = string
+}
+
+############################################################
+# COS variables
+############################################################
+
+variable "cos_guid" {
+  description = "This GUID is the globally unique identifier for the Cloud Object Storage instance."
```
Comment (mkrudele) [inline storage_delegation/variables.tf:16]: Change to "The globally unique identifier of the Cloud Object Storage instance."

File: storage_delegation/variables.tf [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  type        = string
+}
+
+############################################################
+# Key Protect variables
+############################################################
+
+variable "cos_kms_crn" {
+  description = "Key Protect CRN used to encrypt the COS bucket used by the Watson project."
```
Comment (mkrudele) [inline storage_delegation/variables.tf:25]: Change to "Key Protect service instance CRN used to encrypt the COS buckets used by the watsonx projects."

There can be more than one project using the same cos instance.

File: storage_delegation/variables.tf [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    condition = anytrue([
+      can(regex("^crn:(.*:){3}kms:(.*:){2}[0-9a-fA-F]{8}(?:-[0-9a-fA-F]{4}){3}-[0-9a-fA-F]{12}::$", var.cos_kms_crn)),
+      var.cos_kms_crn == null,
+    ])
+    error_message = "Key Protect CRN validation failed."
+  }
+}
+
+variable "cos_kms_key_crn" {
+  description = "Key Protect key CRN used to encrypt the COS bucket used by the Watson project."
```
Comment (mkrudele) [inline storage_delegation/variables.tf:39]: Change to: "Key Protect key CRN used to encrypt the COS buckets used by the watsonx projects. If not set, then the cos_kms_new_key_name must be specified."

File: storage_delegation/variables.tf [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  type        = string
+  default     = null
+}
+
+variable "cos_kms_new_key_name" {
+  description = "Name of the Key Protect key to create to encrypt the COS bucket used by the Watson project."
```
Comment (mkrudele) [inline storage_delegation/variables.tf:45]: Change to "Name of the Key Protect key to create for encrypting the COS buckets used by the watsonx projects."

File: storage_delegation/variables.tf [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  type        = string
+  default     = ""
+}
+
+variable "cos_kms_ring_id" {
+  description = "The ID of the KMS ring to create the cos_kms_new_key_name into. The parameter will only be used if the DA has to create the KMS key. If it is not set, then the new key will be created in the default ring."
```
Comment (mkrudele) [inline storage_delegation/variables.tf:51]: Change to "The identifier of the Key Protect ring to create the cos_kms_new_key_name into. If it is not set, then the new key will be created in the default ring."

File: variables.tf [hunk 6]
```diff
@@ -136,3 +136,35 @@ variable "watsonx_project_tags" {
   type        = list(string)
   default     = ["watsonx-ai-SaaS"]
 }
+
```
Comment (mkrudele) [inline variables.tf:139]: Change the descriptions accordingly with the changes made to the module.

**Review by mkrudele — CHANGES_REQUESTED** [review]

File: configure_project/variables.tf [hunk 7]
```diff
@@ -22,6 +23,11 @@ variable "watsonx_project_tags" {
   type        = list(string)
 }
 
+variable "watsonx_project_delegated" {
+  description = "Set to true of the COS instance is delegated by the account admin."
```
[the remark this answers is not in the document]
Comment (andreainnocenti) [reply configure_project/variables.tf:27]: fixed

File: main.tf [hunk 8]
```diff
@@ -92,6 +92,16 @@ module "configure_user" {
   resource_group_id     = module.resource_group.resource_group_id
 }
 
+module "storage_delegation" {
+  source                = "./storage_delegation"
+  watsonx_admin_api_key = var.watsonx_admin_api_key == null || var.watsonx_admin_api_key == "" ? var.ibmcloud_api_key : var.watsonx_admin_api_key
```
[the remark this answers is not in the document]
Comment (andreainnocenti) [reply main.tf:97]: I used the same current logic: 

if the admin apikey is null or empty -> use the deployer apikey
if the admin apikey is passed -> use the admin apikey

[the remark this answers is not in the document]
Comment (andreainnocenti) [reply main.tf]: fixed

File: storage_delegation/variables.tf [hunk 9]
```diff
@@ -0,0 +1,54 @@
+############################################################
+# Watson variables
+############################################################
+
+variable "watsonx_admin_api_key" {
```
[the remark this answers is not in the document]
Comment (andreainnocenti) [reply storage_delegation/variables.tf]: fixed

**Review by mkrudele — APPROVED** [review]

**Pull request merged**